Added performance tuning guide

jpkrohling opened the pull request:
Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

jpkrohling commented:
@lucperkins, this is using your admonitions feature. Could you check that it looks the way you intended it to?

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,87 @@
+
+### Consider using Apache Kafka as the temporary storage
+
+Jaeger can use Apache Kafka as a buffer between the collector and the actual backing storage (Elasticsearch, Apache Cassandra). This is ideal for cases where the traffic spikes are relatively frequent (prime time traffic) but the storage can eventually catch up once the traffic normalizes. For that, the storage should be set to `kafka` in the collector and the Jaeger Ingester component can be use ...
```
- in the first sentence link to Architecture page
- I'd add "Having spans written to Kafka is useful for building real time data pipeline for aggregations and feature extraction from traces";

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,87 @@
+
+The agent is meant to be placed on the same host as the instrumented application. This is typically accomplished by having one agent per bare metal for traditional applications, or as a sidecar in container environments like Kubernetes, as this helps spread the load handled by agents with the additional advantage of allowing each agent to be tweaked individually, according to the application’s ne ...
+
+### Consider using Apache Kafka as the temporary storage
```
2nd level header, not 3rd. People complained about calling Kafka "storage", so "intermediate buffer".

jpkrohling replied:
It's third-level because it's inside "Deployment architecture".

jpkrohling commented:
I think most of items are addressed, except for moving the Kafka part to 2nd level instead of 3rd.

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+Most of the Jaeger clients, such as the Java, Go, and C#, buffer spans in memory before batching them to the Jaeger Agent/Collector. The maximum size of this buffer is defined by the environment variable `JAEGER_REPORTER_MAX_QUEUE_SIZE` (default value: about `100` spans): the larger the size, the higher the potential memory consumption. When the instrumented application is generating a large numb ...
+
+In the common scenarios, the queue (metric: `jaeger_tracer_reporter_queue_length`) will be close to empty, as spans are flushed at regular intervals to the agent or collector or when the concrete senders, such as the `UdpSender`, deem that a flush is needed.
```
as spans are flushed to the agent or collector at regular intervals or when a certain size of the batch is reached.

jpkrohling replied:
Batch size is an implementation detail of the UdpSender, but I think it's easier to understand while being technically correct.

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+In the common scenarios, the queue (metric: `jaeger_tracer_reporter_queue_length`) will be close to empty, as spans are flushed at regular intervals to the agent or collector or when the concrete senders, such as the `UdpSender`, deem that a flush is needed.
+
+{{< info >}}
```
I suggest not using the `info` highlight here, it's not _that_ important to this doc. Can simpler add it to the last paragraph.

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+### Queue size
+
+Similar to the Agent, the Collector is able to receive spans and place them in an internal queue for processing. This allows the collector to return immediately to the client/agent instead of waiting for the span to make its way to the storage.
```
do we want to always use capitalized works for component names?  We're pretty inconsistent throughout the documentation.

jpkrohling replied:
Good question. I'd vote for keeping the initials capitalized.

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+Jaeger was built from day 1 to be able to ingest huge amounts of data in a resilient way. To better utilize resources that might cause delays, such as storage or network communications, Jaeger buffers and batches data. When more spans are generated than Jaeger is able to safely process, spans might get dropped. However, the defaults might not fit all scenarios: for instance, agents running as a s ...
+
+## Adjusting the client settings
```
For consistency, perhaps use these titles:
  * Client (tracer) settings
  * Agent settings
  * Collector settings (current title)

JStickler replied:
Usually I'd say yes, but in this particular case I think we're OK with having the Level 1 headings that Yuri recommends (the architectural components), as long as we have verbs with the Level 2 headings (which is where you're suggesting some sort of action).
So maybe something like this?  (Feel free to pick different verbs, this is just a quick example)
Client (Tracer) settings
-Adjust sampling configuration
-Increase in-memory queue size
- Modify batched spans flush interval
Agent Settings
- Increase server queue sizes
- Increase processor workers
Collector Settings
- Increase queue size
- Increase workers

jpkrohling replied:
I changed the doc to use your suggestions, except for the last one, which I used "Increase processor workers" (like in the Agent settings).

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+## Adjusting the client settings
+
+### Settings `JAEGER_SAMPLER_TYPE` and `JAEGER_SAMPLER_PARAM`
```
using env names in the heading makes the sidebar menu look rather ugly (one even gets cut off). 

![image](https://user-images.githubusercontent.com/3523016/61314929-b34edb00-a7cb-11e9-885b-c25795fbfcbb.png)


I would recommend

### Sampling configuration

Sampling in the client can be controlled with `JAEGER_SAMPLER_TYPE` and `JAEGER_SAMPLER_PARAM` environment variables. [...]

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+Some clients support the setting `JAEGER_DISABLED` to completely disable the Jaeger Tracer. This is recommended only if the tracer is behaving in a way that causes problems to the instrumented application, as it will not propagate the context to the downstream services.
+{{< /info >}}
+
+### Setting `JAEGER_REPORTER_MAX_QUEUE_SIZE`
```
In-memory queue size (move env var in text)

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+The detailed behavior of this queue is described in this [GitHub issue](https://github.com/jaegertracing/jaeger-client-java/issues/607).
+{{< /info >}}
+
+### Setting `JAEGER_REPORTER_FLUSH_INTERVAL`
```
Batched spans flush interval, env var in text

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+Together, the sampler type and parameter specify how often traces should be "sampled", ie, recorded and sent to the Jaeger backend. For applications generating a large number of spans, setting the sampling type to `probabilistic` and the value to `0.001` (the default) will cause traces to be reported with a 1/1000th chance. Note that the sampling decision is made at the root span and propagated d ...
+
+For applications with low to medium traffic, setting the sampling type to `const` and value to `1` will cause all spans to be reported. Similarly, tracing can be disabled by setting the value to `0`, while context propagation will continue to work.
```
To our internal users we always recommend using remote sampler, so that concrete sampling strategies are decided by the  backend. Even though adaptive sampling isn't live yet in oss, the backend-driven configuration is already supported (../sampling/#collector-sampling-configuration), so perhaps worth mentioning here, as otherwise fixed configurations sound like the recommended approach.

jpkrohling replied:
> We recommend to set your clients to use the [`remote` sampling strategy](./sampling/#collector-sampling-configuration), so that admins can centrally set the concrete sampling strategy for each service.

Like this?

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+For applications with low to medium traffic, setting the sampling type to `const` and value to `1` will cause all spans to be reported. Similarly, tracing can be disabled by setting the value to `0`, while context propagation will continue to work.
+
+{{< info >}}
```
not sure `info` is necessary here, the point is not _that_ important for this doc, and the use of `JAEGER_DISABLED` already makes the text stand out a bit.

jpkrohling replied:
As the text about the remote sampling isn't really about performance, I removed the admonition from this paragraph and added to that new one.

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+### Workers
+
+Items from the span queue in the collector are picked up by workers. Each worker picks one span from the queue and persists it to the storage. The number of workers can be specified by the setting `collector.num-workers`(default: `50`) and should be as high as needed to keep the queue close to zero. The general rule is: the faster the backing storage, the lower the number of workers can be. Given ...
```
missing space before (default: `50`)

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+Items from the span queue in the collector are picked up by workers. Each worker picks one span from the queue and persists it to the storage. The number of workers can be specified by the setting `collector.num-workers`(default: `50`) and should be as high as needed to keep the queue close to zero. The general rule is: the faster the backing storage, the lower the number of workers can be. Given ...
+
+## Deployment architecture
```
suggestion: Deployment considerations

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+### Make sure the storage can keep up
+
+Each span is written to the storage by the collector using one worker, blocking it until the span has been stored. When the storage is too slow, the number of workers blocked by the storage might be too high, causing spans to be dropped. To help diagnose this situation, the histogram `jaeger_collector_save_latency_bucket` can be analyzed. Ideally, this histogram would show a nearly constant value ...
```
> Ideally, this histogram would show a nearly constant value

this sounds odd. A specific percentile can be flat.

jpkrohling replied:
Agree, it's missing an "over time" there. 

> Ideally, this histogram would show a nearly constant value over time

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+### Consider using Apache Kafka as intermediate buffer
+
+Jaeger [can use Apache Kafka](./architecture/) as a buffer between the collector and the actual backing storage (Elasticsearch, Apache Cassandra). This is ideal for cases where the traffic spikes are relatively frequent (prime time traffic) but the storage can eventually catch up once the traffic normalizes. For that, the storage should be set to `kafka` in the collector and the Jaeger Ingester c ...
```
up one level: `(../architecture/)`

yurishkuro replied:
> the storage should be set to `kafka`

the `SPAN_STORAGE_TYPE` should be set to `kafka`

JStickler commented on content/docs/next-release/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+For applications with low to medium traffic, setting the sampling type to `const` and value to `1` will cause all spans to be reported. Similarly, tracing can be disabled by setting the value to `0`, while context propagation will continue to work.
+
+{{< info >}}
+Some clients support the setting `JAEGER_DISABLED` to completely disable the Jaeger Tracer. This is recommended only if the tracer is behaving in a way that causes problems to the instrumented application, as it will not propagate the context to the downstream services.
```
Inconsistent capitalization here.  Is it "tracer" or "Tracer"?  (My vote would be capitalize it)

yurishkuro replied:
yeah, I called it out too. I would capitalize components everywhere (we can have a separate PR to fix it in all other docs)

JStickler commented on content/docs/next-release/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+### Setting `JAEGER_REPORTER_MAX_QUEUE_SIZE`
+
+Most of the Jaeger clients, such as the Java, Go, and C#, buffer spans in memory before batching them to the Jaeger Agent/Collector. The maximum size of this buffer is defined by the environment variable `JAEGER_REPORTER_MAX_QUEUE_SIZE` (default value: about `100` spans): the larger the size, the higher the potential memory consumption. When the instrumented application is generating a large numb ...
```
Either "Most of the Jaeger clients, such as Java, Go, and C#,"
Or
"Most of the Jaeger clients, such as the Java, Go, and C# clients," (my preference)

JStickler commented on content/docs/next-release/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+Most of the Jaeger clients, such as the Java, Go, and C#, buffer spans in memory before batching them to the Jaeger Agent/Collector. The maximum size of this buffer is defined by the environment variable `JAEGER_REPORTER_MAX_QUEUE_SIZE` (default value: about `100` spans): the larger the size, the higher the potential memory consumption. When the instrumented application is generating a large numb ...
+
+In the common scenarios, the queue (metric: `jaeger_tracer_reporter_queue_length`) will be close to empty, as spans are flushed at regular intervals to the agent or collector or when the concrete senders, such as the `UdpSender`, deem that a flush is needed.
```
```suggestion
In most common scenarios, the queue (metric: `jaeger_tracer_reporter_queue_length`) will be close to empty, as spans are flushed at regular intervals to the agent or collector or when the concrete senders, such as the `UdpSender`, deem that a flush is needed.
```

JStickler commented on content/docs/next-release/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+### Make sure the storage can keep up
+
+Each span is written to the storage by the collector using one worker, blocking it until the span has been stored. When the storage is too slow, the number of workers blocked by the storage might be too high, causing spans to be dropped. To help diagnose this situation, the histogram `jaeger_collector_save_latency_bucket` can be analyzed. Ideally, this histogram would show a nearly constant value ...
```
"it’s a good indication that your storage might need some attention."
some attention = more storage capacity?

jpkrohling replied:
Really depends on the concrete instance that is deployed. Perhaps a faster storage is necessary. Perhaps storage needs to be deployed closer to the collector. Perhaps there's another consumer stealing capacity. It's hard to come up with a concrete recommendation here.

JStickler commented on content/docs/next-release/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+Items from the span queue in the collector are picked up by workers. Each worker picks one span from the queue and persists it to the storage. The number of workers can be specified by the setting `collector.num-workers`(default: `50`) and should be as high as needed to keep the queue close to zero. The general rule is: the faster the backing storage, the lower the number of workers can be. Given ...
+
+## Deployment architecture
```
Shouldn't deployment architecture come first?  I especially think that the section about having the agents close to the applications should come before the Reporter_Flush_Interval section where you talk about how the networking overhead affecting that setting.

yurishkuro replied:
good point, +1

jpkrohling commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+### Setting `JAEGER_REPORTER_MAX_QUEUE_SIZE`
+
+Most of the Jaeger clients, such as the Java, Go, and C#, buffer spans in memory before batching them to the Jaeger Agent/Collector. The maximum size of this buffer is defined by the environment variable `JAEGER_REPORTER_MAX_QUEUE_SIZE` (default value: about `100` spans): the larger the size, the higher the potential memory consumption. When the instrumented application is generating a large numb ...
```
Yes, that's how it's printed out from `/metrics`. I think it's even described in the specs :-)

jpkrohling commented:
@JStickler, @yurishkuro thanks for the review! I believe I addressed most of the comments, marking them as "Resolved" as I changed them. I need your feedback in a couple of items, and I left them open.

jpkrohling commented:
About listing the metrics at the end of the section: not sure about that. As a reader, I think I'd prefer to have them close to where it's mentioned ("don't make me think"). If the problem is that it's not visually appealing, we can always summon @lucperkins and have him take a look :)

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,89 @@
+
+### Make sure the storage can keep up
+
+Each span is written to the storage by the Collector using one worker, blocking it until the span has been stored. When the storage is too slow, the number of workers blocked by the storage might be too high, causing spans to be dropped. To help diagnose this situation, the histogram `jaeger_collector_save_latency_bucket` can be analyzed. Ideally, this histogram would show a nearly constant value ...
```
> this histogram would show a nearly constant value over time

Ideally, the latency should remain the same over time.

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,89 @@
+
+### Sampling configuration
+
+Together, the sampler type and parameter specify how often traces should be "sampled", ie, recorded and sent to the Jaeger backend. For applications generating a large number of spans, setting the sampling type to `probabilistic` and the value to `0.001` (the default) will cause traces to be reported with a 1/1000th chance. Note that the sampling decision is made at the root span and propagated d ...
```
Together, the `JAEGER_SAMPLER_TYPE` and `JAEGER_SAMPLER_PARAM` specify

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,89 @@
+
+### In-memory queue size
+
+Most of the Jaeger clients, such as the Java, Go, and C# clients, buffer spans in memory before batching them to the Jaeger Agent/Collector. The maximum size of this buffer is defined by the environment variable `JAEGER_REPORTER_MAX_QUEUE_SIZE` (default value: about `100` spans): the larger the size, the higher the potential memory consumption. When the instrumented application is generating a la ...
```
s/before batching them/before sending them/

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,89 @@
+In addition to the performance aspects, having spans written to Kafka is useful for building real time data pipeline for aggregations and feature extraction from traces.
+
+## Client (Tracer) settings
+
```
I would add a paragraph

"Jaeger Clients can be configured programmatically or via [environment variables] (../client-features/)"

jpkrohling replied:
Added

jpkrohling commented:
Just applied the changes addressing the latest comments. I guess it's now ready to be merged.

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+We recommend to set your clients to use the [`remote` sampling strategy](../sampling/#collector-sampling-configuration), so that admins can centrally set the concrete sampling strategy for each service.
+{{< /info >}}
+
+### Increase in-memory queue size
```
I liked the headers without verbs better. We don’t know that “increase the queue size” is the right action in all cases, maybe the user needs to do the opposite in a particular situation. If we want a verb, we could say “adjust”.

jpkrohling replied:
Replaced.

yurishkuro commented on content/docs/1.13/performance-tuning.md:
```diff
@@ -0,0 +1,91 @@
+
+Some Clients support the setting `JAEGER_DISABLED` to completely disable the Jaeger Tracer. This is recommended only if the Tracer is behaving in a way that causes problems to the instrumented application, as it will not propagate the context to the downstream services.
+
+Jaeger Clients can be configured programmatically or via [environment variables](../client-features/).
```
I meant that it should go as the intro to Client Settings section. It’s a good writing practice not to use two consecutive headers without some text between them.

jpkrohling replied:
Got it, good point. I added a small intro to each section.

jpkrohling commented:
PR squashed and rebased.
